### PR TITLE
[#9] 공통 응답 처리 및 회원가입 기능

### PR DESCRIPTION
**Pull request by JinDDung2**
## 작업 요약

- 공통 응답 및 예외
- 회원가입 기능

## 작업 내용

- 공통 Api 응답
- 공통 예외처리
- 유저 도메인 생성
- 회원가입 기능
- 회원가입 테스트

### CODE REVIEW

**Comment by JinDDung2** [comment]
### 📝 테스트 커버리지 리포트입니다
|File|Coverage [60.61%]|:x:|
|:-|:-:|:-:|
|[UserRole.java](https://github.com/f-lab-edu/givemeticon/blob/dfeb427f9a6034acc33648b6aa6e2386ae8b4b87/src%2Fmain%2Fjava%2Fcom%2Fexample%2Fgivemeticon%2Fuser%2Fdomain%2FUserRole.java)|100%|:green_apple:|
|[UserException.java](https://github.com/f-lab-edu/givemeticon/blob/dfeb427f9a6034acc33648b6aa6e2386ae8b4b87/src%2Fmain%2Fjava%2Fcom%2Fexample%2Fgivemeticon%2Fuser%2Fexception%2FUserException.java)|100%|:green_apple:|
|[DuplicatedPhoneException.java](https://github.com/f-lab-edu/givemeticon/blob/dfeb427f9a6034acc33648b6aa6e2386ae8b4b87/src%2Fmain%2Fjava%2Fcom%2Fexample%2Fgivemeticon%2Fuser%2Fexception%2FDuplicatedPhoneException.java)|100%|:green_apple:|
|[DuplicatedEmailException.java](https://github.com/f-lab-edu/givemeticon/blob/dfeb427f9a6034acc33648b6aa6e2386ae8b4b87/src%2Fmain%2Fjava%2Fcom%2Fexample%2Fgivemeticon%2Fuser%2Fexception%2FDuplicatedEmailException.java)|100%|:green_apple:|
|[ErrorCode.java](https://github.com/f-lab-edu/givemeticon/blob/dfeb427f9a6034acc33648b6aa6e2386ae8b4b87/src%2Fmain%2Fjava%2Fcom%2Fexample%2Fgivemeticon%2Fcommon%2Fexception%2FErrorCode.java)|100%|:green_apple:|
|[UserService.java](https://github.com/f-lab-edu/givemeticon/blob/dfeb427f9a6034acc33648b6aa6e2386ae8b4b87/src%2Fmain%2Fjava%2Fcom%2Fexample%2Fgivemeticon%2Fuser%2Fapplication%2FUserService.java)|100%|:green_apple:|
|[UserController.java](https://github.com/f-lab-edu/givemeticon/blob/dfeb427f9a6034acc33648b6aa6e2386ae8b4b87/src%2Fmain%2Fjava%2Fcom%2Fexample%2Fgivemeticon%2Fuser%2Fpresentation%2FUserController.java)|100%|:green_apple:|
|[PasswordConfig.java](https://github.com/f-lab-edu/givemeticon/blob/dfeb427f9a6034acc33648b6aa6e2386ae8b4b87/src%2Fmain%2Fjava%2Fcom%2Fexample%2Fgivemeticon%2Fuser%2Fconfig%2FPasswordConfig.java)|100%|:green_apple:|
|[SignUpRequest.java](https://github.com/f-lab-edu/givemeticon/blob/dfeb427f9a6034acc33648b6aa6e2386ae8b4b87/src%2Fmain%2Fjava%2Fcom%2Fexample%2Fgivemeticon%2Fuser%2Fpresentation%2Frequest%2FSignUpRequest.java)|58.18%|:x:|
|[User.java](https://github.com/f-lab-edu/givemeticon/blob/dfeb427f9a6034acc33648b6aa6e2386ae8b4b87/src%2Fmain%2Fjava%2Fcom%2Fexample%2Fgivemeticon%2Fuser%2Fdomain%2FUser.java)|54.66%|:x:|
|[GiveMeTiConException.java](https://github.com/f-lab-edu/givemeticon/blob/dfeb427f9a6034acc33648b6aa6e2386ae8b4b87/src%2Fmain%2Fjava%2Fcom%2Fexample%2Fgivemeticon%2Fcommon%2Fexception%2FGiveMeTiConException.java)|44.44%|:x:|
|[UserExceptionAdvice.java](https://github.com/f-lab-edu/givemeticon/blob/dfeb427f9a6034acc33648b6aa6e2386ae8b4b87/src%2Fmain%2Fjava%2Fcom%2Fexample%2Fgivemeticon%2Fuser%2Fadvice%2FUserExceptionAdvice.java)|36.84%|:x:|
|[CommonExceptionAdvice.java](https://github.com/f-lab-edu/givemeticon/blob/dfeb427f9a6034acc33648b6aa6e2386ae8b4b87/src%2Fmain%2Fjava%2Fcom%2Fexample%2Fgivemeticon%2Fcommon%2Fadvice%2FCommonExceptionAdvice.java)|14%|:x:|
|[ApiResponse.java](https://github.com/f-lab-edu/givemeticon/blob/dfeb427f9a6034acc33648b6aa6e2386ae8b4b87/src%2Fmain%2Fjava%2Fcom%2Fexample%2Fgivemeticon%2Fcommon%2Fresponse%2FApiResponse.java)|0%|:x:|
|[ErrorResult.java](https://github.com/f-lab-edu/givemeticon/blob/dfeb427f9a6034acc33648b6aa6e2386ae8b4b87/src%2Fmain%2Fjava%2Fcom%2Fexample%2Fgivemeticon%2Fcommon%2Fresponse%2FErrorResult.java)|0%|:x:|

|Total Project Coverage|60.3%|:x:|
|:-|:-:|:-:|

File: src/main/java/com/example/givemeticon/common/exception/ErrorCode.java [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+
+@Getter
+public enum ErrorCode {
+    // AUTH
+    UNAUTHENTICATED("로그인이 필요한 기능입니다."),
+    UNAUTHORIZED("해당 권한이 없습니다."),
+
+    // MEMBER
+    DUPLICATED_EMAIL("이미 존재하는 이메일입니다."),
+    DUPLICATED_Phone("이미 존재하는 휴대폰 번호입니다."),
```
Comment (f-lab-kyle) [inline src/main/java/com/example/givemeticon/common/exception/ErrorCode.java]:
```suggestion
    DUPLICATED_PHONE("이미 존재하는 휴대폰 번호입니다."),
```

File: src/main/java/com/example/givemeticon/user/application/UserService.java [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+    private final PasswordEncoder passwordEncoder;
+
+    public UserService(UserMapper userMapper, PasswordEncoder passwordEncoder) {
+        this.userMapper = userMapper;
+        this.passwordEncoder = passwordEncoder;
+    }
+
+    public void signUp(SignUpRequest request) {
+        checkUserValidity(request);
+        User user = request.toEntity(passwordEncoder);
```
Comment (f-lab-kyle) [inline src/main/java/com/example/givemeticon/user/application/UserService.java]: PasswordEncoder 객체를 넘겨서 거기서 변환하지 말고 애초에 변환된 문자열을 넘겨 처리하는건 어떨까요?

File: src/main/java/com/example/givemeticon/user/domain/UserRole.java [hunk 3]
```diff
@@ -0,0 +1,11 @@
+package com.example.givemeticon.user.domain;
+
+public enum UserRole {
+    USER("ROLE_USER"),
+    VIP("ROLE_VIP");
```
Comment (f-lab-kyle) [inline src/main/java/com/example/givemeticon/user/domain/UserRole.java]:
```suggestion
    USER, VIP;
```

그냥 이렇게 해도 되지 않을까요?

File: src/main/java/com/example/givemeticon/common/advice/CommonExceptionAdvice.java [hunk 4]
```diff
@@ -0,0 +1,26 @@
+package com.example.givemeticon.common.advice;
```
Comment (f-lab-kyle) [inline src/main/java/com/example/givemeticon/common/advice/CommonExceptionAdvice.java]: example 은 빼도 되지 않을까요?ㅎ

**Review by f-lab-kyle — COMMENTED** [review]
수고하셨습니다~!

File: src/main/java/com/example/givemeticon/user/application/UserService.java (hunk 2 above)
Comment (JinDDung2) [reply src/main/java/com/example/givemeticon/user/application/UserService.java]: String encryptedPassword = passwordEncoder.encode(request.getPassword());
 User user = request.toEntity(encryptedPassword);

위와 같이 바꾸면 다음과 같은 차이가 있습니다.

SRP : 비밀번호 암호화와 사용자 엔티티 생성 및 저장이 명확히 분리되어 있습니다. 각 메서드가 하나의 명확한 역할을 수행하며, 각 역할이 분리되어 있습니다. 
가독성: 한 줄마다 목적이 분명하게 드러나므로 코드를 읽고 이해하기 쉽습니다.

이전 방식은 회원가입처럼 간단한 작업이라 간결한 코드라는 장점이 있고, 바꾼 방식은 SRP와 가독성을 통해 유지보수가 편한 코드인 것 같습니다.

File: src/main/java/com/example/givemeticon/user/domain/UserRole.java (hunk 3 above)
Comment (JinDDung2) [reply src/main/java/com/example/givemeticon/user/domain/UserRole.java]: 다른 블로그들을 보면 "ROLE_"을 붙이는 경우가 많습니다. 이는 스프링 시큐리티가 사용자 역할(권한)을 찾을 때 "ROLE_*"로 인식해서 위와 같은 방식을 사용하는 것으로 알고 있습니다. 
스프링 시큐리티를 사용하지 않는다면 위와 같은 방식은 불필요한 것이기 때문인가요? 혹은 다른 이유가 있나요?

Comment (f-lab-kyle) [reply src/main/java/com/example/givemeticon/user/domain/UserRole.java]: 아, 큰 차이는 없는게 클래스 이름, enum 값에 ROLE 이라는 중복(?) 의미가 있는것 같아서 말씀드린 부분입니다 ㅎ

**Comment by JinDDung2** [comment]
### 📝 테스트 커버리지 리포트입니다
|File|Coverage [62.46%]|:x:|
|:-|:-:|:-:|
|[ErrorCode.java](https://github.com/f-lab-edu/givemeticon/blob/18ee5d63dcd7bbe64fccb8948c7c1d22925413d4/src%2Fmain%2Fjava%2Fcom%2Fjinddung2%2Fgivemeticon%2Fcommon%2Fexception%2FErrorCode.java)|100%|:green_apple:|
|[UserRole.java](https://github.com/f-lab-edu/givemeticon/blob/18ee5d63dcd7bbe64fccb8948c7c1d22925413d4/src%2Fmain%2Fjava%2Fcom%2Fjinddung2%2Fgivemeticon%2Fuser%2Fdomain%2FUserRole.java)|100%|:green_apple:|
|[DuplicatedEmailException.java](https://github.com/f-lab-edu/givemeticon/blob/18ee5d63dcd7bbe64fccb8948c7c1d22925413d4/src%2Fmain%2Fjava%2Fcom%2Fjinddung2%2Fgivemeticon%2Fuser%2Fexception%2FDuplicatedEmailException.java)|100%|:green_apple:|
|[UserException.java](https://github.com/f-lab-edu/givemeticon/blob/18ee5d63dcd7bbe64fccb8948c7c1d22925413d4/src%2Fmain%2Fjava%2Fcom%2Fjinddung2%2Fgivemeticon%2Fuser%2Fexception%2FUserException.java)|100%|:green_apple:|
|[DuplicatedPhoneException.java](https://github.com/f-lab-edu/givemeticon/blob/18ee5d63dcd7bbe64fccb8948c7c1d22925413d4/src%2Fmain%2Fjava%2Fcom%2Fjinddung2%2Fgivemeticon%2Fuser%2Fexception%2FDuplicatedPhoneException.java)|100%|:green_apple:|
|[UserService.java](https://github.com/f-lab-edu/givemeticon/blob/18ee5d63dcd7bbe64fccb8948c7c1d22925413d4/src%2Fmain%2Fjava%2Fcom%2Fjinddung2%2Fgivemeticon%2Fuser%2Fapplication%2FUserService.java)|100%|:green_apple:|
|[UserController.java](https://github.com/f-lab-edu/givemeticon/blob/18ee5d63dcd7bbe64fccb8948c7c1d22925413d4/src%2Fmain%2Fjava%2Fcom%2Fjinddung2%2Fgivemeticon%2Fuser%2Fpresentation%2FUserController.java)|100%|:green_apple:|
|[User.java](https://github.com/f-lab-edu/givemeticon/blob/18ee5d63dcd7bbe64fccb8948c7c1d22925413d4/src%2Fmain%2Fjava%2Fcom%2Fjinddung2%2Fgivemeticon%2Fuser%2Fdomain%2FUser.java)|67.08%|:x:|
|[SignUpRequest.java](https://github.com/f-lab-edu/givemeticon/blob/18ee5d63dcd7bbe64fccb8948c7c1d22925413d4/src%2Fmain%2Fjava%2Fcom%2Fjinddung2%2Fgivemeticon%2Fuser%2Fpresentation%2Frequest%2FSignUpRequest.java)|59.41%|:x:|
|[GiveMeTiConException.java](https://github.com/f-lab-edu/givemeticon/blob/18ee5d63dcd7bbe64fccb8948c7c1d22925413d4/src%2Fmain%2Fjava%2Fcom%2Fjinddung2%2Fgivemeticon%2Fcommon%2Fexception%2FGiveMeTiConException.java)|44.44%|:x:|
|[GivemeticonApplication.java](https://github.com/f-lab-edu/givemeticon/blob/18ee5d63dcd7bbe64fccb8948c7c1d22925413d4/src%2Fmain%2Fjava%2Fcom%2Fjinddung2%2Fgivemeticon%2FGivemeticonApplication.java)|37.5%|:x:|
|[UserExceptionAdvice.java](https://github.com/f-lab-edu/givemeticon/blob/18ee5d63dcd7bbe64fccb8948c7c1d22925413d4/src%2Fmain%2Fjava%2Fcom%2Fjinddung2%2Fgivemeticon%2Fuser%2Fadvice%2FUserExceptionAdvice.java)|36.84%|:x:|
|[CommonExceptionAdvice.java](https://github.com/f-lab-edu/givemeticon/blob/18ee5d63dcd7bbe64fccb8948c7c1d22925413d4/src%2Fmain%2Fjava%2Fcom%2Fjinddung2%2Fgivemeticon%2Fcommon%2Fadvice%2FCommonExceptionAdvice.java)|14%|:x:|
|[ErrorResult.java](https://github.com/f-lab-edu/givemeticon/blob/18ee5d63dcd7bbe64fccb8948c7c1d22925413d4/src%2Fmain%2Fjava%2Fcom%2Fjinddung2%2Fgivemeticon%2Fcommon%2Fresponse%2FErrorResult.java)|0%|:x:|
|[ApiResponse.java](https://github.com/f-lab-edu/givemeticon/blob/18ee5d63dcd7bbe64fccb8948c7c1d22925413d4/src%2Fmain%2Fjava%2Fcom%2Fjinddung2%2Fgivemeticon%2Fcommon%2Fresponse%2FApiResponse.java)|0%|:x:|
|[PasswordConfig.java](https://github.com/f-lab-edu/givemeticon/blob/18ee5d63dcd7bbe64fccb8948c7c1d22925413d4/src%2Fmain%2Fjava%2Fcom%2Fjinddung2%2Fgivemeticon%2Fuser%2Fconfig%2FPasswordConfig.java)|0%|:x:|

|Total Project Coverage|62.46%|:x:|
|:-|:-:|:-:|

**Pull request merged**